[Grid][aspect-ratio] Block size computed from aspect ratio is definite for grid layout.

sammygill opened the pull request:
#### 30864f828c216943fd2ec5272f053f45757f7ffe
<pre>
[Grid][aspect-ratio] Block size computed from aspect ratio is definite for grid layout.
<a href="https://bugs.webkit.org/show_bug.cgi?id=282764">https://bugs.webkit.org/show_bug.cgi?id=282764</a>
<a href="https://rdar.apple.com/139444773">rdar://139444773</a>

Reviewed by Brent Fulgham.

If a grid computes its block size from its inline size and aspect ratio, then that block
size should be treated as definite for the purposes of grid layout. In the attached test
case, the grid computes its inline size from the available space and its resulting block
size from that and the 1/1 aspect ratio.

While the grid ends up getting sized correctly in the end, we fail to take into
consideration this definite size for grid layout, which could have a result of not properly
aligning items correctly. This can be seen in the state-of-html-2024 survey within a button
to delete series from filters.

Let&apos;s fix this by expanding hasDefiniteLogicalHeight to consider
logical heights computed from the aspect ratio. Later on during grid layout, we will recognize
this, compute a logical height from the aspect ratio, and use it for the purposes of
row sizing.

Aside: In theory, we should be able to resolve this by hoisting up updateLogicalWidth
to right before we compute hasDefiniteLogicalHeight, and it should get resolved in
RenderBlock::hasDefiniteLogicalHeight, which already has logic to handle this scenario.
Sadly, this does not quite work, however, because that logic ends up bailing out early if the
grid&apos;s preferredLogicalWidths bits are dirty even if we do not use them for layout.

* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/aspect-ratio/grid-aspect-ratio-align-items-center-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/aspect-ratio/grid-aspect-ratio-align-items-center.html: Added.
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::layoutGrid):

Canonical link: <a href="https://commits.webkit.org/286367@main">https://commits.webkit.org/286367@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/184d4d17de2e6647421697e4a3ecb88d18b35383

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/75756 "Passed style check") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/54785 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28608 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80245 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27022 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/77872 "Passed tests") | [  ~~🛠 ios-sim~~](https://ews-build.webkit.org/#/builders/49/builds/63927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3045 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59412 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17580 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/78823 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/49/builds/63927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65073 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39768 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/49/builds/63927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22555 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25349 "Built successfully") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/49/builds/63927 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/22894 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81716 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3096 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/1964 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67643 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3249 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65040 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/66945 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16694 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/10893 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/9031 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3054 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/5880 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3079 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/4015 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/3086 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

brentfulgham reviewed (approved):
Nice WPT progression. r=me

sammygill committed:
[Grid][aspect-ratio] Block size computed from aspect ratio is definite for grid layout.
https://bugs.webkit.org/show_bug.cgi?id=282764
rdar://139444773

Reviewed by Brent Fulgham.

If a grid computes its block size from its inline size and aspect ratio, then that block
size should be treated as definite for the purposes of grid layout. In the attached test
case, the grid computes its inline size from the available space and its resulting block
size from that and the 1/1 aspect ratio.

While the grid ends up getting sized correctly in the end, we fail to take into
consideration this definite size for grid layout, which could have a result of not properly
aligning items correctly. This can be seen in the state-of-html-2024 survey within a button
to delete series from filters.

Let's fix this by expanding hasDefiniteLogicalHeight to consider
logical heights computed from the aspect ratio. Later on during grid layout, we will recognize
this, compute a logical height from the aspect ratio, and use it for the purposes of
row sizing.

Aside: In theory, we should be able to resolve this by hoisting up updateLogicalWidth
to right before we compute hasDefiniteLogicalHeight, and it should get resolved in
RenderBlock::hasDefiniteLogicalHeight, which already has logic to handle this scenario.
Sadly, this does not quite work, however, because that logic ends up bailing out early if the
grid's preferredLogicalWidths bits are dirty even if we do not use them for layout.

* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/aspect-ratio/grid-aspect-ratio-align-items-center-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-sizing/aspect-ratio/grid-aspect-ratio-align-items-center.html: Added.
* Source/WebCore/rendering/RenderGrid.cpp:
(WebCore::RenderGrid::layoutGrid):

Canonical link: https://commits.webkit.org/286367@main

alanbaradlay commented on Source/WebCore/rendering/RenderGrid.cpp:
```diff
@@ -328,7 +328,7 @@ void RenderGrid::layoutGrid(bool relayoutChildren)
         // FIXME: Consider caching the hasDefiniteLogicalHeight value throughout the layout.
         // FIXME: We might need to cache the hasDefiniteLogicalHeight if the call of RenderBlock::hasDefiniteLogicalHeight() causes a relevant performance regression.
-        bool hasDefiniteLogicalHeight = RenderBlock::hasDefiniteLogicalHeight() || overridingLogicalHeight() || computeContentLogicalHeight(RenderBox::SizeType::MainOrPreferredSize, style().logicalHeight(), std::nullopt);
+        bool hasDefiniteLogicalHeight = RenderBlock::hasDefiniteLogicalHeight() || overridingLogicalHeight() || computeContentLogicalHeight(RenderBox::SizeType::MainOrPreferredSize, style().logicalHeight(), std::nullopt) || shouldComputeLogicalHeightFromAspectRatio();
```
I think this may need a comment as it is not super clear why hasDefiniteLogicalHeight is supposed to be true when shouldComputeLogicalHeightFromAspectRatio() is true (while it is pretty obvious when I look at  the previous values like RenderBlock::hasDefiniteLogicalHeight()).
